### PR TITLE
Binding for `try_for` tactic

### PR DESCRIPTION
**Pull request by tyleretzel**
https://github.com/prove-rs/z3.rs/issues/154

### CODE REVIEW

File: z3/tests/lib.rs [hunk 1]
```diff
@@ -1128,6 +1128,30 @@ fn test_tactic_skip() {
...
+    let two = ast::Int::from_i64(&ctx, 2);
+    let x = ast::Int::new_const(&ctx, "x");
+
+    let goal = Goal::new(&ctx, false, false, false);
+    goal.assert(&x.ge(&one.add(&two)));
+
+    let tactic = Tactic::new(&ctx, "simplify");
+    let try_for_tactic = tactic.try_for(4294967295);
+
+    // Test that the `try_for` can successfully apply underlying tactic
```
Comment (tyleretzel) [inline z3/tests/lib.rs:1147]: It'd be great to have a tactic that reliably uses more than `n` milliseconds to execute (then we could test the failure case here), but I couldn't find anything that fits this bill

File: z3/src/tactic.rs [hunk 2]
```diff
@@ -103,6 +103,20 @@ impl<'ctx> Tactic<'ctx> {
         }
     }
 
+    /// Return a tactic that applies the current tactic to the goal, failing if
+    /// `t` doesn't terminate in `timeout_ms` milliseconds.
+    pub fn try_for(&self, timeout_ms: u32) -> Tactic<'ctx> {
```
Comment (sameer) [inline z3/src/tactic.rs:108]: Could we use [Duration](https://doc.rust-lang.org/std/time/struct.Duration.html) here?

File: z3/tests/lib.rs [hunk 3]
```diff
@@ -1128,6 +1128,30 @@ fn test_tactic_skip() {
...
+
+    let one = ast::Int::from_i64(&ctx, 1);
+    let two = ast::Int::from_i64(&ctx, 2);
+    let x = ast::Int::new_const(&ctx, "x");
+
+    let goal = Goal::new(&ctx, false, false, false);
+    goal.assert(&x.ge(&one.add(&two)));
+
+    let tactic = Tactic::new(&ctx, "simplify");
+    let try_for_tactic = tactic.try_for(4294967295);
```
Comment (sameer) [inline z3/tests/lib.rs:1145]: Please use the associated const [`u32::MAX`](https://doc.rust-lang.org/std/primitive.u32.html#associatedconstant.MAX) for this

File: z3/tests/lib.rs (hunk 1 above)
Comment (sameer) [reply z3/tests/lib.rs:1147]: I'm not aware of a tactic like that, but testing success should be good enough

**Review by sameer — CHANGES_REQUESTED** [review]
Thanks for submitting this! Just had a few comments -- after that it should be good to go

**Comment by tyleretzel** [comment]
I did some more testing and tried to actually use this, and I think there's some more fundamental issues with the tactics API: it doesn't seem like we handle tactic "failure" correctly (it always causes a segfault). Even something simple like this will trigger it:

```
#[test]
fn test_tactic_fail() {
    let cfg = Config::new();
    let ctx = Context::new(&cfg);
    let params = Params::new(&ctx);

    let a = ast::Bool::new_const(&ctx, "a");
    let goal = Goal::new(&ctx, false, false, false);
    goal.assert(&a);

    let tactic = Tactic::new(&ctx, "fail");
    let apply_results = tactic.apply(&goal, Some(&params));
    let goal_results = apply_results.list_subgoals().collect::<Vec<Goal>>();
    let goal_result = goal_results.first().unwrap();
    assert_eq!(goal_result.get_formulas::<Bool>(), vec![a],);
}
```

Let me try to see what's going on here and fix this, too.

**Comment by sameer** [comment]
Good find, keep me posted! :slightly_smiling_face:

**Pull request closed without merging**